hazelcast-sql/src/test/java/com/hazelcast/sql/impl/client/SqlResubmissionTest.java

```diff
@@ -0,0 +1,228 @@
+
+    @Before
+    public void initFailure() {
+        clusterFailure.initialize(INITIAL_CLUSTER_SIZE, smallInstanceConfig());
+        createMap(clusterFailure.getNotFailingInstance(), COMMON_MAP_NAME, COMMON_MAP_SIZE, IntHolder::new,
+                IntHolder.class);
+        ClientConfig clientConfig = new ClientConfig();
+        clientConfig.getSqlConfig().setSqlResubmissionMode(resubmissionMode);
+
+        client = clusterFailure.createClient(clientConfig);
+        state = State.BEFORE_EXECUTE;
+        done = false;
+        lastExecutionTime = 0;
+    }
+
+    @Test
+    public void when_failingSelectBeforeAnyDataIsFetched() throws InterruptedException {
+        SqlStatement statement = new SqlStatement("select * from " + COMMON_MAP_NAME);
+        testStatement(statement, shouldFailBeforeAnyDataIsFetched(resubmissionMode));
+    }
+
+    @Test
+    public void when_failingUpdate() throws InterruptedException {
+        SqlStatement statement = new SqlStatement("update " + COMMON_MAP_NAME + " set field = 1");
+        testStatement(statement, shouldFailNonSelectQuery(resubmissionMode));
+    }
+
+    private void testStatement(SqlStatement statement, boolean shouldFail)
+            throws InterruptedException {
+        Thread failingThread = new Thread(cyclicFailure);
+        failingThread.start();
+        try {
+            if (shouldFail) {
+                assertThrows(HazelcastSqlException.class,
+                        () -> executeInLoop(statement, result -> false));
+                done = true;
+            } else {
+                try {
+                    executeInLoop(statement, result -> ((SqlClientResult) result).wasResubmission());
+                } finally {
+                    done = true;
+                }
+            }
+        } finally {
+            failingThread.join();
+            clusterFailure.cleanUp();
+        }
+    }
+
+    private void executeInLoop(SqlStatement statement, Predicate<SqlResult> shouldBreakFunction) {
+        while (!done) {
+            State localState = state;
+            if (localState == State.BEFORE_EXECUTE) {
+                state = State.BEFORE_FAIL;
+                try {
+                    long start = System.nanoTime();
+                    SqlResult result = client.getSql().execute(statement);
+                    lastExecutionTime = (System.nanoTime() - start) / 1_000_000;
+                    if (shouldBreakFunction.test(result)) {
+                        break;
+                    }
+                } catch (RuntimeException e) {
+                    if (e.getMessage() != null && e.getMessage().contains("CREATE MAPPING")) {
+                        continue;
+                    }
+                    throw e;
+                }
+            }
+        }
+    }
+
+    private final Runnable cyclicFailure = () -> {
+        boolean failureNeeded = true;
+
+        while (!done) {
+            State localState = state;
+            switch (localState) {
```

Comment:
This code spins the CPU at 100% when in `BEFORE_EXECUTE` - shouldn't we sleep a little?

Reply:
The state ```BEFORE_EXECUTE``` last very short. It switches to ```BEFORE_FAIL``` in on instruction after other thread sees that state, that was my motivation to spin:

```java
            State localState = state;
            if (localState == State.BEFORE_EXECUTE) {
                state = State.BEFORE_FAIL;
```